### PR TITLE
Core: Add base implementations for changelog tasks

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds implementations for changelog tasks to be used in the changelog scan.
This PR currently relies on existing tests. More tests will come with the scan implementation.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
@flyrain @rdblue @stevenzwu @karuppayya @szehon-ho @RussellSpitzer @anuragmantri

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:55]: There seems to be some ways to have a method toStringHelper() at each level and use it for inheritance, do you think its worth it?  https://github.com/google/guava/issues/1239

File: core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java:74]: Is it possible in  parent class to have abstract construction method:      

protected abstract ThisT newSplitScanTask(T parentTask, int start, int offset);

and then push merge() to parent, using this construction method?  Then we get merge() and canMerge() in the same class.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:61]: ThisT and T are a bit confusing, how about something like T => ParentT ?

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:26]: nit: `ThisT` seems a little uncommon compared to just `T`

Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:26]: does `ThisT` refers to this class of `BaseChangelogContentScanTask`?

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java:39]: do we need to validate?

1. offsetList is in strictly ascending order
2. last offset is smaller than the parentTaskLength

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java:37]: nit: is `splitIndex` more accurate?

File: core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java:74]: I didn't quite follow the merge semantics here. we just add the length but lose track of `other` in the returned object.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java:74]: @szehon-ho, I am not sure that will be any cleaner as the only purpose of `merge` would be to directly call that new abstract method. I feel like it would pretty much the same.

@stevenzwu, that's the existing logic from `FileScanTask`. If two tasks are adjacent after split planning, we merge them together into one.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:55]: The string representation would include the actual class name. I thought that would be sufficient. I don't think we include the file path or location of deletes files right now. The rest is the same across changelog tasks.

Locations may have some sensitive data. I'd probably keep it as-is for now to match the existing behavior. We can definitely reconsider the behavior in the future.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:61]: I agree. I was trying to stay on one line as the list of params is long but let me try to find some reasonable formatting to use `ParentT`.

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java:37]: I copied this code as it was. From what I see, it looks like an index to the size of this iterator that is used to access both offsets and split sizes. Let me do some renames to make it a bit cleaner.

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java:39]: I am a bit worried to add/change the existing behavior in this class. I know we had a bug related to point 2 when the size of the file in metadata was incorrect. I guess handling point 1 may be a good idea. Let me think a bit more.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:61]: @szehon-ho, instead of making the base class abstract and extending it everywhere, I can probably accept a closure to construct a split scan task in the base class. Then we won't need any subclasses. I think that's going to be cleaner. What do you think?

Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:61]: That sounds great

File: core/src/main/java/org/apache/iceberg/BaseContentScanTask.java
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/BaseContentScanTask.java:28]: Nit: Does it make sense to use `T extends ScanTask` instead of `ThisT`? So that the type of `ThisT` is more obvious here.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java]: I managed to implement it. Could you take a look, @szehon-ho?

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java]: Changed.

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java]: I added validation for sorted offsets, which is required by the spec.

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java]: I changed the code in `next` a little bit so hopefully the purpose of `sizeIndex` is a bit clearer now.

File: core/src/main/java/org/apache/iceberg/BaseContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseContentScanTask.java]: We are using `ThisT` throughout the code to have access to the actual type of the object in a parent class. Usually, we don't annotate it with boundaries. However, I had to in this case. 

It became this:

```
ThisT extends ContentScanTask<F>
```

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (stevenzwu) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java]: >  it is an index of the size of this iterator that is used to access both offsets and split sizes lists

That is exactly where I found it a little confusing to be named as `sizeIndex`. This index is used to construct a split from the indexed offsets and sizes lists.

```
    long offset = offsets.get(sizeIndex);
    long splitSize = splitSizes.get(sizeIndex);
    sizeIndex += 1; // create 1 split per offset
    return createSplitTaskFunc.apply(parentTask, offset, splitSize);
```

But this is a very minor thing.

**Review by stevenzwu — APPROVED** [review]
This looks quite good to me

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java]: Yeah, I see what you mean now. I don't know what was the original interpretation but I can switch to `splitIndex`.

File: core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java:32]: Minor naming suggestion:
```
 @FunctionalInterface
  interface SplitTaskCreator<T extends ScanTask> {
  T create(T parentTask, long offset, long length);
  }
```

**Review by flyrain — APPROVED** [review]
LGTM

File: core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java]: Yea its not reducing a lot of code, just more symmetric for the reader.  The classes below (OffsetsAwareSplitScanTaskIteratorImpl and FixedSizeSplitScanTaskIteratorImpl) already have this 'newSplitTask' method, so we would just have one more just like that.

```
  private static class SplitScanTaskImpl
      extends SplitScanTask<SplitScanTaskImpl, AddedRowsScanTask, DataFile>
      implements AddedRowsScanTask {

    SplitScanTaskImpl(AddedRowsScanTask parentTask, long offset, long length) {
      super(parentTask, offset, length);
    }

    @Override
    public List<DeleteFile> deletes() {
      return parentTask().deletes();
    }

    @Override
    protected SplitScanTaskImpl newSplitTask(AddedRowsScanTask parentTask, long offset, long length) {
      return new SplitScanTaskImpl(parentTask, offset, length);
    }
  }
```

So the merge() is not just directly calling the newSplitTask but actually doing somewhat corresponding logic to canMerge(), which now is more clear in the same class.

```
  abstract static class SplitScanTask<ThisT, T extends ContentScanTask<F> & ChangelogScanTask, F extends ContentFile<F>>
      implements ContentScanTask<F>, ChangelogScanTask, MergeableScanTask<ThisT> {
   ...
       abstract ThisT newSplitTask(T parentTask, long offset, long length);
    
    @Override
    public boolean canMerge(ScanTask other) {
      if (getClass().equals(other.getClass())) {
        SplitScanTask<?, ?, ?> that = (SplitScanTask<?, ?, ?>) other;
        return changeOrdinal() == that.changeOrdinal() &&
            commitSnapshotId() == that.commitSnapshotId() &&
            file().equals(that.file()) &&
            start() + length() == that.start();

      } else {
        return false;
      }
    }

    @Override
    public ThisT merge(ScanTask other) {
      SplitScanTask that = (SplitScanTask) other;
      return newSplitTask(parentTask(), start(), length() + that.length());
    }
   ...
}
```

If we can somehow get rid of all these redundant classes with a lambda or something that would be the best, but probably its not possible.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:56]: Hm , maybe I might have been not clear, so just to clarify it wasn't suggesting to make toString in the child clases and logging other fields like location, but rather to reduce the code by having a method toStringHelper() from the base class (BaseContentScanTask) and re-use it (where we already have log file, partition_data, and residual) ?

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:56]: Oh, okay, got it now. Let me explore this idea.

File: core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java]: Okay, I've changed the approach to have `merge` in the parent. Let me know if that's what you meant.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:56]: I tried it locally. I couldn't control the order of fields in child classes and I couldn't benefit from it in split scan tasks as that one does not extend `BaseContentScanTask`. I'd just probably keep it as-is.

File: core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java:32]: Missed it. That's a good idea. Let me change.

File: core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/OffsetsAwareSplitScanTaskIterator.java]: Switched.

File: core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SplitScanTaskIterator.java]: Done.

File: core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java]: Yea, I didn't realize that you would remove the newSplitTask from the bottom two methods, leaving this one to be the only one.  But it's better to me.

File: core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java:56]: Sounds good.

**Review by szehon-ho — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @flyrain @stevenzwu @szehon-ho!

I merged this to unblock the scan implementation. If anyone has any additional comments, I'll be happy to follow up.

**Pull request merged**
